test/nat46x64/test.sh

```diff
@@ -1,423 +1,411 @@
-2.2.2.2/15 0 1.1.1.1/18 80 TCP,\
-2.2.2.2/16 0 1.1.1.1/17 80 TCP,\
-2.2.2.2/17 0 1.1.1.1/16 80 TCP,\
-2.2.2.2/18 0 1.1.1.1/15 80 TCP,\
-2.2.2.2/19 0 1.1.1.1/14 80 TCP,\
-2.2.2.2/20 0 1.1.1.1/13 80 TCP,\
-2.2.2.2/21 0 1.1.1.1/12 80 TCP,\
-2.2.2.2/22 0 1.1.1.1/11 80 TCP,\
-2.2.2.2/23 0 1.1.1.1/10 80 TCP,\
-2.2.2.2/24 0 1.1.1.1/9 80 TCP,\
-2.2.2.2/25 0 1.1.1.1/8 80 TCP,\
-2.2.2.2/26 0 1.1.1.1/7 80 TCP,\
-2.2.2.2/27 0 1.1.1.1/6 80 TCP,\
-2.2.2.2/28 0 1.1.1.1/5 80 TCP,\
-2.2.2.2/29 0 1.1.1.1/4 80 TCP,\
-2.2.2.2/30 0 1.1.1.1/3 80 TCP,\
-2.2.2.2/31 0 1.1.1.1/2 80 TCP,\
-2.2.2.2/32 0 1.1.1.1/1 80 TCP,\
-2.2.2.2/32 0 1.1.1.1/0 80 TCP"
+        --filters="$(printf '%s,' "${RECORDER_FILTERS_IPV4[@]}" | sed 's/,*$//')"
 
 # Trigger recompilation with 32 IPv6 filter masks
 ${CILIUM_EXEC} \
     cilium-dbg recorder update --id 2 --caplen 100 \
-        --filters="f00d::1/0 80 cafe::/128 0 UDP,\
-f00d::1/1 80 cafe::/127 0 UDP,\
-f00d::1/2 80 cafe::/126 0 UDP,\
-f00d::1/3 80 cafe::/125 0 UDP,\
-f00d::1/4 80 cafe::/124 0 UDP,\
-f00d::1/5 80 cafe::/123 0 UDP,\
-f00d::1/6 80 cafe::/122 0 UDP,\
-f00d::1/7 80 cafe::/121 0 UDP,\
-f00d::1/8 80 cafe::/120 0 UDP,\
-f00d::1/9 80 cafe::/119 0 UDP,\
-f00d::1/10 80 cafe::/118 0 UDP,\
-f00d::1/11 80 cafe::/117 0 UDP,\
-f00d::1/12 80 cafe::/116 0 UDP,\
-f00d::1/13 80 cafe::/115 0 UDP,\
-f00d::1/14 80 cafe::/114 0 UDP,\
-f00d::1/15 80 cafe::/113 0 UDP,\
-f00d::1/16 80 cafe::/112 0 UDP,\
-f00d::1/17 80 cafe::/111 0 UDP,\
-f00d::1/18 80 cafe::/110 0 UDP,\
-f00d::1/19 80 cafe::/109 0 UDP,\
-f00d::1/20 80 cafe::/108 0 UDP,\
-f00d::1/21 80 cafe::/107 0 UDP,\
-f00d::1/22 80 cafe::/106 0 UDP,\
-f00d::1/23 80 cafe::/105 0 UDP,\
-f00d::1/24 80 cafe::/104 0 UDP,\
-f00d::1/25 80 cafe::/103 0 UDP,\
-f00d::1/26 80 cafe::/102 0 UDP,\
-f00d::1/27 80 cafe::/101 0 UDP,\
-f00d::1/28 80 cafe::/100 0 UDP,\
-f00d::1/29 80 cafe::/99 0 UDP,\
-f00d::1/30 80 cafe::/98 0 UDP,\
-f00d::1/31 80 cafe::/97 0 UDP,\
-f00d::1/32 80 cafe::/96 0 UDP,\
-f00d::1/32 80 cafe::/0 0 UDP"
-
-${CILIUM_EXEC} cilium-dbg recorder list
-${CILIUM_EXEC} cilium-dbg bpf recorder list
+        --filters="$(printf '%s,' "${RECORDER_FILTERS_IPV6[@]}" | sed 's/,*$//')"
+
+info "Checking the list of filters in IPv4 recorder ID:1"
+if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:1" | wc -l) -ne ${#RECORDER_FILTERS_IPV4[@]} ]; then
+    echo "Expected filters:"
+    echo "${RECORDER_FILTERS_IPV4[@]}" | nl -bn
+    echo "Found filters:"
+    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+    fatal "Recorder filters did not match expected list"
+fi
+
+info "Checking the list of filters in IPv6 recorder ID:2"
+if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:2" | wc -l) -ne ${#RECORDER_FILTERS_IPV6[@]} ]; then
+    echo "Expected filters:"
+    echo "${RECORDER_FILTERS_IPV6[@]}" | nl -bn
+    echo "Found filters:"
+    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+    fatal "Recorder filters did not match expected list"
+fi
```

Comment:
nit: these two blocks look similar enough that they might be refactored into a function as well.

Reply:
I agree. I did try this earlier when developing these tests, but being able to properly pass the array contents into a function was a little tricky. I can try to take another stab at this, but if it's not trivial then I'll probably skip this feedback. IMO we're reaching diminishing returns on iterating on these steps at this point.

Recorder is also a bit weird to be combined into the nat46x64 test, it seems to be entirely unrelated; I wouldn't mind if it were in its own file. But a bit more refactoring is necessary to get that done. We can continue to iterate in the tree.

Reply:
Looks easy enough. I'll add an extra commit on top:

```
diff --git a/test/nat46x64/test.sh b/test/nat46x64/test.sh
index 6366f448a838..75a5aa0af4e0 100755
--- a/test/nat46x64/test.sh
+++ b/test/nat46x64/test.sh
@@ -258,6 +258,23 @@ function test_services {
     ${CILIUM_EXEC} cilium-dbg service delete 2
 }

+# Check whether the list of filters matches the expected list.
+#
+# $1 - ID
+# $@ - Filters
+function check_recorder_list {
+    ID=$1
+    shift
+    num_filters=$#
+    if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:${ID}" | wc -l) -ne "$num_filters" ]; then
+        echo "Expected filters:"
+        echo "$@" | nl -bn
+        echo "Found filters:"
+        ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+        fatal "Recorder filters did not match expected list"
+    fi
+}
+
 force_cleanup 2>&1 >/dev/null
 initialize_docker_env
 trap cleanup EXIT
@@ -383,23 +400,8 @@ ${CILIUM_EXEC} \
     cilium-dbg recorder update --id 2 --caplen 100 \
         --filters="$(printf '%s,' "${RECORDER_FILTERS_IPV6[@]}" | sed 's/,*$//')"

-info "Checking the list of filters in IPv4 recorder ID:1"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:1" | wc -l) -ne ${#RECORDER_FILTERS_IPV4[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV4[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
-
-info "Checking the list of filters in IPv6 recorder ID:2"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:2" | wc -l) -ne ${#RECORDER_FILTERS_IPV6[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV6[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
+check_recorder_list 1 "${RECORDER_FILTERS_IPV4[@]}"
+check_recorder_list 2 "${RECORDER_FILTERS_IPV6[@]}"

 trace_exec ${CILIUM_EXEC} cilium-dbg recorder list
 trace_exec ${CILIUM_EXEC} cilium-dbg bpf recorder list
```